content/blog/2018/egress-mongo/index.md

```diff
@@ -0,0 +1,1120 @@
+    apiVersion: networking.istio.io/v1alpha3
+    kind: ServiceEntry
+    metadata:
+      name: mongo
+    spec:
+      hosts:
+      - "*.com"
+      ports:
+      - number: 443
+        name: tls
+        protocol: TLS
+      - number: $MONGODB_PORT
+        name: tls-mongodb
+        protocol: TLS
+      location: MESH_EXTERNAL
+    EOF
+    {{< /text >}}
+
+1.  Create an egress `Gateway` for _*.com_, port 443, protocol TLS, a destination rule to set the
+    [SNI](https://en.wikipedia.org/wiki/Server_Name_Indication) for the gateway, and a virtual service to direct the
+    traffic destined for _*.com_ to the gateway.
+
+    {{< text bash >}}
+    $ kubectl apply -f - <<EOF
+    apiVersion: networking.istio.io/v1alpha3
+    kind: Gateway
+    metadata:
+      name: istio-egressgateway-with-sni-proxy
+    spec:
+      selector:
+        istio: egressgateway-with-sni-proxy
+      servers:
+      - port:
+          number: 443
+          name: tls
+          protocol: TLS
+        hosts:
+        - "*.com"
+        tls:
+          mode: MUTUAL
+          serverCertificate: /etc/certs/cert-chain.pem
+          privateKey: /etc/certs/key.pem
+          caCertificates: /etc/certs/root-cert.pem
+    ---
+    apiVersion: networking.istio.io/v1alpha3
+    kind: DestinationRule
+    metadata:
+      name: set-sni-for-egress-gateway
+    spec:
+      host: istio-egressgateway-with-sni-proxy.istio-system.svc.cluster.local
+      subsets:
+        - name: mongo
+          trafficPolicy:
+            loadBalancer:
+              simple: ROUND_ROBIN
+            portLevelSettings:
+            - port:
+                number: 443
+              tls:
+                mode: ISTIO_MUTUAL
+                sni: placeholder.com
+    EOF
+    {{< /text >}}
+
+1.  Route the traffic destined for _*.com_ to the egress gateway and from the egress gateway to the SNI proxy.
+
+    {{< text bash >}}
+    $ kubectl apply -f - <<EOF
+    apiVersion: networking.istio.io/v1alpha3
+    kind: VirtualService
+    metadata:
+      name: direct-mongo-through-egress-gateway
+    spec:
+      hosts:
+      - "*.com"
+      gateways:
+      - mesh
+      - istio-egressgateway-with-sni-proxy
+      tls:
+      - match:
+        - gateways:
+          - mesh
+          port: $MONGODB_PORT
+          sni_hosts:
+          - "*.com"
+        route:
+        - destination:
+            host: istio-egressgateway-with-sni-proxy.istio-system.svc.cluster.local
+            subset: mongo
+            port:
+              number: 443
+          weight: 100
+      tcp:
+      - match:
+        - gateways:
+          - istio-egressgateway-with-sni-proxy
+          port: 443
+        route:
+        - destination:
+            host: sni-proxy.local
+            port:
+              number: $MONGODB_PORT
+          weight: 100
+    EOF
+    {{< /text >}}
+
+1.  Refresh the web page of the application again and verify that the ratings are still displayed correctly.
```

Comment:
I'm missing something here. How is the application using this config for `*.com`? Isn't the ratings service using `MONGO_DB_URL=mongodb://bookinfo:$BOOKINFO_PASSWORD@$MONGODB_HOST:$MONGODB_PORT/test?authSource=test&ssl=true` for it's requests?

Reply:
@frankbu Good point - the application does not really use multiple mongodb hosts. The usecase is applicable when the cluster has to access multiple MongoDB instances, with urls for example, like: mongodb1.composedb.com, mongodb2.composedb.com. Instead of specifying a configuration per each MongoDB host, the user can specify a single configuration for *.composedb.com. Since the application uses a single MongoDB host, it would be possible to specify exactly that mongodb host, as it was done in the previous section.

This section shows that you can specify `*.composedb.com` to allow the traffic to `mongodb1.composedb.com`. The reader can deploy a second mongodb host, e.g. `mongodb2.composedb.com`, let bookinfo use it and see that everything works with the second host without any reconfiguration, since the wildcard was used. I can propose it in the blog post as an exercise for the reader.

What do you think?

Reply:
Explaining this and leaving it as an exercise for the user sounds good. I think you also need to make clear that `*.com` works for your compose, become the mongodb host you are using ends in `.com` (e.g., `mongodb1.composedb.com`). If someone is trying this out with some other mongodb instance, that ends in `.org` for example, they would need to use something like `*.org` for the config instead. I think you just need to explain what's going on a little better

Reply:
Sounds good, please see https://github.com/istio/istio.io/pull/2347/commits/723e7be8a7a86c40d6362ce7a5c860ef27f39674